cassandraprotocol/client/client_test.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+var compressors = map[string]*frame.Codec{
+	"LZ4":    frame.NewCodec(frame.WithCompressor(&compression.Lz4Compressor{})),
+	"SNAPPY": frame.NewCodec(frame.WithCompressor(&compression.SnappyCompressor{})),
+	"NONE":   frame.NewCodec(),
+}
+
+// This test requires a remote server listening on localhost:9042 without authentication.
+func TestRemoteServerNoAuth(t *testing.T) {
+
+	for version := cassandraprotocol.ProtocolVersionMin; version <= cassandraprotocol.ProtocolVersionMax; version++ {
+		t.Run(fmt.Sprintf("version %v", version), func(t *testing.T) {
+
+			for compressor, frameCodec := range compressors {
+				t.Run(fmt.Sprintf("compression %v", compressor), func(t *testing.T) {
+
+					client := NewCqlClient("127.0.0.1:9042", frameCodec)
+					clientConn, err := client.Connect()
+					if err != nil {
+						return // skip test, no remote server available
+					}
+					defer func() { _ = clientConn.Close() }()
+
+					err = Handshake(clientConn, version, 1)
+					assert.Nil(t, err)
+
+					request, _ := frame.NewRequestFrame(version, 1, false, nil, &message.Query{
+						Query:   "SELECT * FROM system.local",
+						Options: message.NewQueryOptions(),
+					})
+					err = clientConn.Send(request)
+					assert.Nil(t, err)
+					fmt.Printf("CLIENT sent:     %v\n", request)
+
+					response, err := clientConn.Receive()
+					assert.Nil(t, err)
+					assert.NotNil(t, response)
+					assert.IsType(t, &message.RowsResult{}, response.Body.Message)
+					fmt.Printf("CLIENT received: %v\n", response)
+				})
+			}
+		})
+	}
+}
+
+// This test requires a remote server listening on localhost:9042 with authentication.
+func TestRemoteServerAuth(t *testing.T) {
+
+	for version := cassandraprotocol.ProtocolVersionMin; version <= cassandraprotocol.ProtocolVersionMax; version++ {
+		t.Run(fmt.Sprintf("version %v", version), func(t *testing.T) {
+
+			for compressor, frameCodec := range compressors {
+				t.Run(fmt.Sprintf("compression %v", compressor), func(t *testing.T) {
+
+					client := NewCqlClient("127.0.0.1:9042", frameCodec)
+					clientConn, err := client.Connect()
+					if err != nil {
+						return // skip test, no remote server available
+					}
+					defer func() { _ = clientConn.Close() }()
+
+					err = HandshakeAuth(clientConn, version, 1, "cassandra", "cassandra")
+					assert.Nil(t, err)
+
+					query, _ := frame.NewRequestFrame(version, 1, false, nil, &message.Query{
+						Query:   "SELECT * FROM system.local",
+						Options: message.NewQueryOptions(),
+					})
+					err = clientConn.Send(query)
+					assert.Nil(t, err)
+					fmt.Printf("CLIENT sent:     %v\n", query)
+
+					rows, err := clientConn.Receive()
+					assert.Nil(t, err)
+					assert.NotNil(t, rows)
+					assert.IsType(t, &message.RowsResult{}, rows.Body.Message)
+					fmt.Printf("CLIENT received: %v\n", rows)
+				})
+			}
+		})
+	}
+}
+
+func TestLocalServer(t *testing.T) {
```

Comment:
This test and the following one create a ping-pong game between a client and a server.